Introduce first class support for Android

couven92 opened the pull request:
Adds detection and support for Android and ARMv71 to the niminst `build.sh`.

Adds necessary when statements for the android POSIX stat function

Most notably, this PR introduces the ability to override the tempDir path using a compile-time define option. (https://github.com/couven92/Nim/commit/849791946488622dde55de7813eecfa2f2f77bfa)
ref. https://github.com/nim-lang/Nim/issues/5260

However, as @Araq points out in https://github.com/nim-lang/Nim/issues/5260#issuecomment-274257315 we should really introduce Android as a first class OS. I have therefore introduced changes to the platform definitions in both the compile and the stdlib and modified the default configuration.

That enables us to do the following in the default `nim.cfg`
``` nimconfig
@if android:
  passL:"-landroid-glob"
  define:"tempDir:~/../tmp"
@end
```

couven92 commented:
Hmm... the trick with simply adding Android to platforms.nim didn't really work though...
``` nim
import os

echo system.hostOS
echo os.getTempDir

#[
Produces on android: nim c -d:android -r test.nim
linux
/tmp/
]#
```

But I really don't want to undefine `linux` just to be able to auto-detect `android`...

However, is it a good idea to move [the line for android in platforms.nim](https://github.com/couven92/Nim/blob/69614346ceb23f8a4363dbb2206d550c39c2042c/lib/system/platforms.nim#L60) up and above the [linux line](https://github.com/couven92/Nim/blob/69614346ceb23f8a4363dbb2206d550c39c2042c/lib/system/platforms.nim#L43) so that android is detected before linux? @Araq

TiberiumN commented:
Yay! But add a note that this support is usable only in termux, not in plain android

couven92 commented:
@TiberiumN you could probably do this in the adb shell as well... If you somehow magically got gcc and android-glob there...

TiberiumN commented:
@couven92 yep :) what's why I mean :)
but anyway, good PR, please merge

yglukhov commented:
@couven92 you seem to be pairing `defined(android)` with `defined(macosx)` which looks wrong because android is much closer to linux than macosx. I might be wrong, just 2 cents.

couven92 commented:
@yglukhov This actually comes from @stisa who came up with this as posted it here: https://github.com/nim-lang/Nim/pull/5646#issuecomment-290931873

yglukhov commented:
Ok, just checked against android headers and these definitions are indeed correct.

yglukhov commented:
Does `--os:android` now imply `-d:android` and `-d:linux`?

yglukhov commented on config/nim.cfg:
```diff
@@ -55,7 +55,8 @@ path="$lib/pure"
   stacktrace:off
-  linetrace:off
+
+linetrace:off
```
Looks like a typo

yglukhov commented on lib/pure/ospaths.nim:
```diff
@@ -523,6 +523,11 @@ when declared(getEnv) or defined(nimscript):
     tags: [ReadEnvEffect, ReadIOEffect].} =
     ## Returns the temporary directory of the current user for applications to
     ## save temporary files in.
+    when defined(tempDir):
```
Whats the reason for this? Should not` getEnv("TMPDIR")` work?

couven92 commented:
@yglukhov unfortunenately this does not work... But yeah, I'd like android to also define linux...

But even `nim compile --os:android -d:linux koch` results in the following error:
> lib/system/dyncalls.nim(161, 10) Error: no implementation for dyncalls

And that is because it does not hit the first when branch that asks for `linux` to be defined... Something is very wrong here... :(

couven92 commented:
I also noticed that when running `uname -o` from the adb shell i get
> Linux

as the name of the OS. It's only termux that is so nice to give return `Android` to us when running `uname -o` (needed for `build.sh`)

dom96 commented:
Can we fix the tests and resolve conflicts with devel?

I also think that the tempDir define needs to be documented somewhere (together with the android target). The ``nimc`` document perhaps?

couven92 commented:
@dom96 Merged with `devel` (15876f6) and added `tempDir` documentation (8b80991) as requested.

Let's see if the tests pass...

I'd welcome additional ideas on how to deal with the Android is Linux regarding the compiler and stdlib platform definitions. I'm not really sure how to best do that...

But for now this branch and the `csources` it generates do compile Nim on Android using Termux.
After nim is compiled, the following settings need to be set in `nim.cfg` (I recommend `~/.config/nim.cfg`)
``` nim.cfg
define:android
passL:"-landroid-glob"
define:"tempDir:$HOME/../tmp"
```

Since we do not actutually have `gcc` (or sth. similar) in the adb-shell, I have not yet been able to make this work directly on the device without Termux.

dom96 commented:
Unfortunately tests are still failing.

On 16 May 2017 at 15:48, Fredrik Høisæther Rasch <notifications@github.com>
wrote:

> @dom96 <https://github.com/dom96> Merged with devel (15876f6
> <https://github.com/nim-lang/Nim/commit/15876f643992872b0f2fc9a870c8caa7c6730155>)
> and added tempDir documentation (8b80991
> <https://github.com/nim-lang/Nim/commit/8b809914a89cfbcdaf0a81d158cf0fe5481d6d3a>)
> as requested.
>
> Let's see if the tests pass...
>
> I'd welcome additional ideas on how to deal with the Android is Linux
> regarding the compiler and stdlib platform definitions. I'm not really sure
> how to best do that...
>
> But for now this branch and the csources it generates do compile Nim on
> Android using Termux.
> After nim is compiled, the following settings need to be set in nim.cfg
> (I recommend ~/.config/nim.cfg)
>
> define:android
> passL:"-landroid-glob"
> define:"tempDir:$HOME/../tmp"
>
> Since we do not actutually have gcc (or sth. similar) in the adb-shell, I
> have not yet been able to make this work directly on the device without
> Termux.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nim-lang/Nim/pull/5772#issuecomment-301806297>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAPDe6fd8AMgjM6oyf-pPjMUDMBufgCnks5r6bctgaJpZM4NPD06>
> .
>

couven92 commented:
Note 76571a9f28ef3cd0b9bdfc511c7202d9658c9e20 is yet another ugly hack to stop test from failing because they do not find `sh`, but are ahppy to run `/bin/sh`. I introduced another `use` define switch for that, but maybe something else should be done here... 

Note that I updated https://github.com/nim-lang/Nim/pull/5772#issuecomment-301806297 to reflect these changes for the configuration on Android

Araq commented on lib/pure/ospaths.nim:
```diff
@@ -532,6 +532,11 @@ when declared(getEnv) or defined(nimscript):
     ## save temporary files in.
+    when defined(tempDir):
+      const tempDir {.strdefine.}: string = nil
+      if not tempDir.isNil() and tempDir.len > 0:
```
This ``if`` is unnecessary.

couven92 replied:
What if someone does `--define:useShPath:` (i.e. sets the sh path to the empty string)? Their own fault? I would be happy about that :P

Araq replied:
Yeah, his own fault.

Araq commented on lib/pure/osproc.nim:
```diff
@@ -763,6 +763,10 @@ elif not defined(useNimRtl):
     if poEvalCommand in options:
       sysCommand = "/bin/sh"
+      when defined(useShPath):
+        var shPath {.strdefine.}: string = nil
```
This is convoluted.

```nim
const shPath {.strdefine.} = "/bin/sh"
```

couven92 replied:
@Araq, thanks for the suggestion did exactly that :) 3141048b53f63e7789c829a56bfee3113e46f60e

TiberiumN commented:
Any progress here?

couven92 commented:
@TiberiumN I have put some thought into it, but had to work on other projects lately. But mostly, I am not really sure on how to best proceed with the problems I have pointed out here. And mostly this is because I do not know the inner workings of the Nim compiler well enough with regards to the areas that will be affected (i.e. I don't know where to look and where to change things).

Any help will be very much appreciated.

Araq commented:
The problem I see is that the Android cannot be detected reliably in ``build.sh`` but maybe we can have a ``build_android.sh`` that calls build.sh with OS=android as a workaround.

couven92 commented:
@Araq, that would be possible and probably a good idea as this would also give the oppurtunity to create the necessary config file as I outlined above (https://github.com/nim-lang/Nim/pull/5772#issuecomment-301806297).

What abut the platform definitions in the stdlib? Were commits 6e609fedee1341bfae8f394d5d6e82cfd1827f86 and 178541349ac6f9339efd0a40a37fbde95023f6c7 the correct way to introduce Android as a platform? Is there sth missing there?, because those changes did not seem to work? 

How do I also include Linux when Android is defined? I assume that I will have to extend the definition here: https://github.com/nim-lang/Nim/pull/5772/commits/6e609fedee1341bfae8f394d5d6e82cfd1827f86#diff-d48e5407c641a83758acd2f13f84e6afR60

couven92 commented:
Hmm, okay it seems we now got a stable build pipeline that works seemlessly on my own HTC One M8 Android phone...

However, I now get this from nimble:
```
-bash-4.4$ nimble install c2nim --verbose
    Reading official package list
Downloading https://github.com/nim-lang/c2nim using git
     Error: Specified directory does not contain a .nimble file.
-bash-4.4$ ls -lah ~/../tmp
total 12
drwx------    3 u0_a147  u0_a147     4.0K Jun 21 17:33 .
drwx------    5 u0_a147  u0_a147     4.0K Jun 21 17:25 ..
drwx------    3 u0_a147  u0_a147     4.0K Jun 21 17:33 nimble_27846
-bash-4.4$ ls -lah ~/../tmp/nimble_27846/
total 12
drwx------    3 u0_a147  u0_a147     4.0K Jun 21 17:33 .
drwx------    3 u0_a147  u0_a147     4.0K Jun 21 17:33 ..
drwx------    5 u0_a147  u0_a147     4.0K Jun 21 17:33 githubcom_nimlangc2nim
-bash-4.4$ ls -lah ~/../tmp/nimble_27846/githubcom_nimlangc2nim/
total 184
drwx------    5 u0_a147  u0_a147     4.0K Jun 21 17:33 .
drwx------    3 u0_a147  u0_a147     4.0K Jun 21 17:33 ..
drwx------    7 u0_a147  u0_a147     4.0K Jun 21 17:33 .git
-rw-------    1 u0_a147  u0_a147       65 Jun 21 17:33 .gitignore
-rw-------    1 u0_a147  u0_a147      422 Jun 21 17:33 .travis.yml
-rw-------    1 u0_a147  u0_a147     1.1K Jun 21 17:33 LICENSE.txt
-rw-------    1 u0_a147  u0_a147      885 Jun 21 17:33 README.md
-rw-------    1 u0_a147  u0_a147     4.8K Jun 21 17:33 c2nim.nim
-rw-------    1 u0_a147  u0_a147      336 Jun 21 17:33 c2nim.nimble
-rw-------    1 u0_a147  u0_a147    23.3K Jun 21 17:33 clex.nim
-rw-------    1 u0_a147  u0_a147    84.0K Jun 21 17:33 cparse.nim
-rw-------    1 u0_a147  u0_a147    14.4K Jun 21 17:33 cpp.nim
drwx------    2 u0_a147  u0_a147     4.0K Jun 21 17:33 doc
-rw-------    1 u0_a147  u0_a147     3.6K Jun 21 17:33 postprocessor.nim
-rw-------    1 u0_a147  u0_a147     5.0K Jun 21 17:33 rules.nim
drwx------    4 u0_a147  u0_a147     4.0K Jun 21 17:33 testsuite
```
*Note: nimble uses `~/../tmp` as temporary directory as `/tmp` is not accessible on Android an the `nim.cfg` overrides the tmp path.*

couven92 commented:
In addtion to the nimble failure, I have now completed a test run with testament on my Android phone and posted the results at http://thnetii.td.org.uit.no/testament-results

Short summary: 1500 tests, 136 failures (~9%)

This will at least give us a rough roadmap of what needs to be done to get all tests to pass, I have noticed some tests fail, because the tests are bad, or because I hadn't yet installed NodeJS on my phone. But some tests might be real issues.

couven92 commented:
After installing nodejs on my Android and re-running the JS category tests, we're now down to 50/1500 failing tests on Android (3,33%)!

v3ss0n commented:
very interesting work you guys doing.

TiberiumN commented:
Any progress on this?

couven92 commented:
**UPDATE**: After some changes and discussion with @Araq the missing tmp directory issue on Android has now been solved by using `~/nimtempfs/` as in `getTempDir` on Android.

Current Andoird Testament Test Run results on a ARMv7 HTC One m8 phone can be found here: http://thnetii.td.org.uit.no/testament-results.html

Currently it shows 24 / 1478 tests failing, which is about 1.62%, so we're slowly getting there.

@Araq, csources generation from this branch are stable and consistently generate compilable sources for Nim on Android. Nim also correctly understands the `android` OS as a `--os` value and imitates `linux` for all cases except those where `when defined(android)` are used explicitly.

I'd suggest that this branch is ready to be merged now.

couven92 commented:
Okay, now we also got working ARMv8 (i.e. arm64) support for Android (thanks for testing that @TiberiumN)

I have also identified problems with getaddrinfo on android, the issue being that Android behaves like old OpenBSD when it comes to networking and as a result IPv6 support is lacking. (ref.: https://stackoverflow.com/questions/39674147/getaddrinfo-on-android-returning-error-eai-badflags )

With ARMv7 (the platform I am testing on) being 32-bit we also uncovered #6189 

With these issue addressed, we are now down to 19 / 1478 failing tests, so it's getting better and better :)

Some of these fails are still related to the Boehm GC, which according to @Araq we can ignore as we don't have boehm installed on android.

couven92 commented:
@zah, since #5770 I've marked you as our concepts genius. :P  
Do you have insights into the failures within the concepts category? http://thnetii.td.org.uit.no/testament-results.html#panel-testResult-1714

couven92 commented:
@TiberiumN was kind enough to lend us some processing time on his ARMv8 phone, i.e. an ARM64 platform. His test results can be found here: http://thnetii.td.org.uit.no/testament-results-TiberiumN.html

We can see 18 / 1332 (1.29%) tests failing (JS targed tests were stripped since he didn't have nodejs installed)
